### PR TITLE
Close target connection upon endpoint timeout

### PR DESCRIPTION
**Pull request by SanojPunchihewa**
## Purpose
In the current architecture, when the endpoint timeouts the timeout handler will call the fault sequence and finish the mediation flow for that request. But the connection to the backend is not closed until the socket timeout.

This fix will improve the architecture to close the socket connection when an endpoint timeout happens for HTTP transport.

Fixes: https://github.com/wso2/api-manager/issues/1746

### CODE REVIEW

File: modules/core/src/main/java/org/apache/synapse/core/axis2/TimeoutHandler.java [hunk 1]
```diff
@@ -233,6 +233,8 @@ private void processCallbacks() {
                     if (RuntimeStatisticCollector.isStatisticsEnabled()) {
                         CallbackStatisticCollector.callbackCompletionEvent(callback.getSynapseOutMsgCtx(), (String) key);
                     }
+                    // Call the TransportSender's onAppError method to release any resources
+                    callback.getAxis2OutMsgCtx().getTransportOut().getSender().onAppError(callback.getAxis2OutMsgCtx());
                 }
```
Comment (isudana) [inline modules/core/src/main/java/org/apache/synapse/core/axis2/TimeoutHandler.java:238]: Can there be a NPE? if sender is not available for some reason? shall we handle those cases?

Comment (SanojPunchihewa) [reply modules/core/src/main/java/org/apache/synapse/core/axis2/TimeoutHandler.java:238]: The sending flow will break if the sender is not available and won't reach this. As a safety measure, I'll add a null check

Comment (SanojPunchihewa) [reply modules/core/src/main/java/org/apache/synapse/core/axis2/TimeoutHandler.java:242]: Fixed with 0328c2c

**Review by isudana — APPROVED** [review]

**Comment by SanojPunchihewa** [comment]
Closing this PR, since the changes are added via https://github.com/wso2/wso2-synapse/pull/2076

**Pull request closed without merging**